src/services/fraud-prover.service.ts

```diff
@@ -0,0 +1,936 @@
+
+    this.state.l1Provider = new L1ProviderWrapper(
+      this.options.l1RpcProvider,
+      this.state.OVM_StateCommitmentChain,
+      this.state.OVM_CanonicalTransactionChain
+    )
+
+    this.state.nextUnverifiedStateRoot =
+      this.options.fromL2TransactionIndex || 0
+  }
+
+  protected async _start(): Promise<void> {
+    while (this.running) {
+      await sleep(this.options.pollingInterval)
+
+      try {
+        this.logger.info(`Looking for mismatched state roots...`)
+        const fraudulentStateRootIndex = await this._findNextFraudulentStateRoot()
+
+        if (fraudulentStateRootIndex === undefined) {
+          this.logger.info(
+            `Didn't find any mismatched state roots. Trying again in ${Math.floor(
+              this.options.pollingInterval / 1000
+            )} seconds...`
+          )
+          continue
+        }
+
+        this.logger.interesting(
+          `Found a mismatched state root at index: ${fraudulentStateRootIndex}`
+        )
+
+        this.logger.info(`Pulling fraud proof data...`)
+        const proof = await this._getFraudProofData(fraudulentStateRootIndex)
+
+        this.logger.info(`Initializing the fraud verification process...`)
+        await this._initializeFraudVerification(
+          proof.preStateRootProof,
+          proof.transactionProof
+        )
+
+        this.logger.info(`Loading fraud proof contracts...`)
+        const {
+          OVM_StateTransitioner,
+          OVM_StateManager,
+        } = await this._getFraudProofContracts(
+          await this.state.l1Provider.getStateRoot(
+            fraudulentStateRootIndex - 1
+          ),
+          proof.transactionProof.transaction
+        )
+
+        // PRE_EXECUTION phase.
+        if (
+          (await OVM_StateTransitioner.phase()) ===
+          StateTransitionPhase.PRE_EXECUTION
+        ) {
+          try {
+            this.logger.interesting(
+              `Fraud proof is now in the PRE_EXECUTION phase.`
+            )
+
+            this.logger.interesting(`Proving account states...`)
+            await this._proveAccountStates(
+              OVM_StateTransitioner,
+              OVM_StateManager,
+              proof.stateDiffProof.accountStateProofs,
+              fraudulentStateRootIndex
+            )
+
+            this.logger.interesting(`Proving storage slot states...`)
+            await this._proveContractStorageStates(
+              OVM_StateTransitioner,
+              proof.stateDiffProof.accountStateProofs
+            )
+
+            this.logger.interesting(`Executing transaction...`)
+            await OVM_StateTransitioner.applyTransaction(
+              proof.transactionProof.transaction,
+              {
+                gasLimit: this.options.runGasLimit,
+              }
+            )
+
+            this.logger.success(`Transaction successfully executed.`)
+          } catch (err) {
+            if (
+              err
+                .toString()
```

Comment:
This seems brittle to a refactor in the contracts

Reply:
I think this is ok but we just need to remember in the future that there is a hard dependency on the require statements in the contracts because this code isn't covered with tests

Reply:
Yeahhhh not sure how to handle this otherwise. My logic has to depend on revert reasons.